### PR TITLE
Check user permissions, attach user during pipeline creation

### PR DESCRIPTION
**Pull request by tkyi**
This pull request will check if a user has admin access to a given repo and run pipeline create if they do; requires `getPermissions` to be added to Models first

Waiting on https://github.com/screwdriver-cd/models/pull/23

### CODE REVIEW

File: test/plugins/pipelines.test.js [hunk 1]
```diff
@@ -16,8 +16,16 @@ sinon.assert.expose(assert, { prefix: '' });
...
 function pipelineModelFactoryMock() {}
 
+/**
+ * Stub for UserModel factory method
+ * @method userModelFactoryMock
+ */
+function userModelFactoryMock() {}
+
 describe('pipeline plugin test', () => {
+    // let hoekMock;
```
Comment (d2lam) [inline test/plugins/pipelines.test.js]: commented code?

File: test/plugins/pipelines.test.js [hunk 2]
```diff
@@ -29,6 +37,9 @@ describe('pipeline plugin test', () => {
     });
 
     beforeEach((done) => {
+        // hoekMock = {
```
Comment (d2lam) [inline test/plugins/pipelines.test.js]: here too

File: test/plugins/pipelines.test.js [hunk 3]
```diff
@@ -226,6 +252,12 @@ describe('pipeline plugin test', () => {
         const dateNow = 1111111111;
         const scmUrl = 'git@github.com:screwdriver-cd/data-model.git#master';
         const testId = 'd398fb192747c9a0124e9e5b4e6e8e841cf8c71c';
+        const username = 'd2lam';
+        const pipeline = {
+            id: testId,
+            other: 'dataToBeIncluded'
+        };
+        // const password = 'this_is_a_password_that_needs_to_be_atleast_32_characters';
```
Comment (d2lam) [inline test/plugins/pipelines.test.js]: heree

File: plugins/pipelines/create.js [hunk 4]
```diff
@@ -18,38 +20,49 @@ module.exports = (datastore) => ({
...
-                    Pipeline.sync({
-                        scmUrl
-                    }, (e) => {
-                        if (e) {
-                            return reply(boom.wrap(e));
-                        }
+                    return Pipeline.create(pipelineConfig, next);
+                },
+                (pipeline, next) => Pipeline.sync({ scmUrl }, next)
+            /* eslint-disable consistent-return */
```
Comment (d2lam) [inline plugins/pipelines/create.js]: do we still need this?

File: plugins/pipelines/create.js [hunk 5]
```diff
@@ -18,38 +20,53 @@ module.exports = (datastore) => ({
...
-                        host: request.headers.host,
-                        port: request.headers.port,
-                        protocol: request.server.info.protocol,
-                        pathname: `${request.path}/${result.id}`
-                    });
+                    return Pipeline.get(pipelineId, next);
+                },
+                (data, next) => {
+                    if (data) {
+                        next(new Error('scmUrl needs to be unique'));
```
Comment (d2lam) [inline plugins/pipelines/create.js]: Can do `return reply(boom.conflict...))` here actually

File: plugins/pipelines/create.js [hunk 6]
```diff
@@ -18,38 +20,53 @@ module.exports = (datastore) => ({
...
+                    }
+                    const pipelineConfig = hoek.merge(request.payload,
+                        {
+                            admins:
+                            {
+                                username: true
```
Comment (d2lam) [inline plugins/pipelines/create.js]: This might set the key as just the string 'username' instead of the actual value... I'm not sure.

Comment (d2lam) [inline plugins/pipelines/create.js]:
```
const admins = {};
admins[username] = true;
const pipelineConfig = hoek.merge(request.payload, admins)
```

File: plugins/pipelines/create.js [hunk 7]
```diff
@@ -18,38 +20,53 @@ module.exports = (datastore) => ({
...
+                    return reply(boom.wrap(err));
+                }
+                const location = urlLib.format({
+                    host: request.headers.host,
+                    port: request.headers.port,
+                    protocol: request.server.info.protocol,
+                    pathname: `${request.path}/${result.id}`
                 });
+
+                return reply(result).header('Location', location).code(201);
```
Comment (d2lam) [inline plugins/pipelines/create.js]: This `result` would be the `main` job instead of the new pipeline because sync returns the job.

Comment (d2lam) [inline plugins/pipelines/create.js]: Nice. Fixed! 👍

File: plugins/pipelines/create.js [hunk 8]
```diff
@@ -18,38 +20,51 @@ module.exports = (datastore) => ({
...
+                    const admins = {};
 
-                    Pipeline.sync({
-                        scmUrl
-                    }, (e) => {
-                        if (e) {
-                            return reply(boom.wrap(e));
-                        }
+                    admins[username] = true;
+                    const pipelineConfig = hoek.merge(request.payload, { admins });
```
Comment (nkatzman) [inline plugins/pipelines/create.js]: should we `hoek.applyToDefaults` so we don't merge values into the `request.payload` ?

Comment (d2lam) [inline plugins/pipelines/create.js]: True

**Comment by d2lam** [comment]
LGTM 👍

**Comment by nkatzman** [comment]
Looks good 👍

**Pull request merged**